fedn/fedn/network/clients/client.py

```diff
@@ -709,14 +703,18 @@ def _send_heartbeat(self, update_frequency=2.0):
             except grpc.RpcError as e:
                 status_code = e.code()
                 if status_code == grpc.StatusCode.UNAVAILABLE:
-                    logger.warning("GRPC hearbeat: server unavailable during send heartbeat. Retrying.")
+                    self._missed_heartbeat += 1
+                    logger.error("GRPC hearbeat: combiner unavailable, retrying (attempt {}/{}).".format(self._missed_heartbeat,
+                                                                                                         self.config['reconnect_after_missed_heartbeat']))
+                    if self._missed_heartbeat > self.config['reconnect_after_missed_heartbeat']:
+                        self.disconnect()
                 if status_code == grpc.StatusCode.UNAUTHENTICATED:
                     details = e.details()
                     if details == 'Token expired':
-                        logger.warning("GRPC hearbeat: Token expired. Reconnecting.")
-                        self.detach()
+                        logger.error("GRPC hearbeat: Token expired. Disconnecting.")
+                        self.disconnect()
+                        sys.exit("Unauthorized. Token expired. Please obtain a new token.")
```

Comment:
Why? this will break the client trying to refresh token?

Reply:
Yes, was thinking about that, but currently the client is not trying to refresh the token, right? So for now the right things seems to be to disconnect?

Reply:
no that's not right they will try to refresh tokens, if you ask for new assignment and the token has expired it will try to get a new one. But only when calling REST-API